src/com/facebook/buck/ide/intellij/BaseIjModuleRule.java

```diff
@@ -118,18 +134,87 @@ private void addDepsAndFolder(
     }
   }
 
+  private void addDepsAndFolder(
+      IJFolderFactory folderFactory,
+      DependencyType dependencyType,
+      TargetNode<T, ?> targetNode,
+      boolean wantsPackagePrefix,
+      ModuleBuildContext context) {
+    addDepsAndFolderWithFolderInputIndexFilter(
+        folderFactory, dependencyType, targetNode, wantsPackagePrefix, context, null);
+  }
+
+  protected void addDepsAndSourcesWithFolderInputIndexFilter(
+      TargetNode<T, ?> targetNode,
+      boolean wantsPackagePrefix,
+      ModuleBuildContext context,
+      @Nullable Predicate<? super Map.Entry<Path, Path>> folderInputIndexFilter) {
+    addDepsAndFolderWithFolderInputIndexFilter(
+        SourceFolder.FACTORY,
+        DependencyType.PROD,
+        targetNode,
+        wantsPackagePrefix,
+        context,
+        folderInputIndexFilter);
+  }
+
   protected void addDepsAndSources(
       TargetNode<T, ?> targetNode, boolean wantsPackagePrefix, ModuleBuildContext context) {
     addDepsAndFolder(
         SourceFolder.FACTORY, DependencyType.PROD, targetNode, wantsPackagePrefix, context);
   }
 
+  protected void addDepsAndTestSourcesWithFolderInputIndexFilter(
+      TargetNode<T, ?> targetNode,
+      boolean wantsPackagePrefix,
+      ModuleBuildContext context,
+      @Nullable Predicate<? super Map.Entry<Path, Path>> folderInputIndexFilter) {
+    addDepsAndFolderWithFolderInputIndexFilter(
+        TestFolder.FACTORY,
+        DependencyType.TEST,
+        targetNode,
+        wantsPackagePrefix,
+        context,
+        folderInputIndexFilter);
+  }
+
   protected void addDepsAndTestSources(
       TargetNode<T, ?> targetNode, boolean wantsPackagePrefix, ModuleBuildContext context) {
     addDepsAndFolder(
         TestFolder.FACTORY, DependencyType.TEST, targetNode, wantsPackagePrefix, context);
   }
 
+  // If there is no resources_root, then we use the java src_roots option from .buckconfig, so
+  // the default folders generated should work. On the other hand, if there is a resources_root,
+  // then for resources under this root, we need to create java-resource folders with the
+  // correct relativeOutputPath set. We also return a filter that removes the resources that we've
+  // added.
+  protected Predicate<Map.Entry<Path, Path>> addResourcesAndGetFilter(
+      ResourceFolderType resourceFolderType,
+      Collection<SourcePath> resources,
+      Path resourcesRoot,
+      ModuleBuildContext context) {
+    ImmutableSet<Path> resourcePaths =
+        resources
+            .stream()
+            .map(
+                sourcePath ->
+                    (sourcePath instanceof PathSourcePath)
+                        ? ((PathSourcePath) sourcePath).getRelativePath()
+                        : null)
+            .filter(Objects::nonNull)
+            .filter(path -> path.startsWith(resourcesRoot))
+            .collect(MoreCollectors.toImmutableSet());
+
+    addSourceFolders(
```

Comment:
This is incorrect, it's adding resource folders, not source folders. I recommend adding a separate interface:

```
public interface ResourceFolderFactory {
  IjFolder create(Path path, Path resourceRoot, ImmutableSortedSet<Path> inputs);
}
```

Then add `addResourceFolders` method:

```
addResourceFolders(
  resourceFolderType.getFactory(),
  getSourceFoldersToInputsIndex(resourcePaths),
  resourcesRoot,
  context);
```